dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java

```diff
@@ -0,0 +1,234 @@
+            this.scheduler = () -> value;
+            return this;
+        }
+
+        @CheckReturnValue
+        public DialogueChannel build() {
+            ClientConfiguration conf = Preconditions.checkNotNull(config, "ClientConfiguration is required");
+            preconditions(channels, conf);
+
+            VersionedTaggedMetricRegistry taggedMetrics =
+                    new VersionedTaggedMetricRegistry(conf.taggedMetricRegistry());
+            DialogueClientMetrics clientMetrics = DialogueClientMetrics.of(taggedMetrics);
+
+            List<LimitedChannel> limitedChannels = channels.stream()
+                    .map(chan -> {
+                        LimitedChannel firstInstance = buildSingleUri(chan, clientMetrics, conf, clock);
+                        LimitedChannel secondInstance = buildSingleUri(chan, clientMetrics, conf, clock);
+                        Preconditions.checkState(
+                                firstInstance.equals(secondInstance),
+                                "a good equals implementation is necessary for smart live-reloading",
+                                UnsafeArg.of("first", firstInstance),
+                                UnsafeArg.of("second", secondInstance));
+                        return firstInstance; // fine to throw away the secondInstance
+                    })
+                    .collect(ImmutableList.toImmutableList());
+
+            LimitedChannel nodeSelectionStrategy =
+                    nodeSelectionStrategy(limitedChannels, taggedMetrics, conf, maybeExisting, random);
+            Channel channel = new LimitedChannelToChannelAdapter(nodeSelectionStrategy);
+            channel = new TracedChannel(channel, "Dialogue-request-attempt");
+            channel = retryingChannel(channel, conf, scheduler, random);
+            channel = new UserAgentChannel(channel, conf.userAgent().get());
+            channel = new DeprecationWarningChannel(channel, clientMetrics);
+            channel = new ContentDecodingChannel(channel);
+            channel = new NeverThrowChannel(channel);
+            channel = new TracedChannel(channel, "Dialogue-request");
+
+            return new DialogueChannel(channel, nodeSelectionStrategy);
+        }
+
+        private static LimitedChannel buildSingleUri(
+                Channel channel, DialogueClientMetrics clientMetrics, ClientConfiguration conf, Ticker clock) {
+            // Instrument inner-most channel with metrics so that we measure only the over-the-wire-time
+            Channel chan = new InstrumentedChannel(channel, clientMetrics);
+            // TracedChannel must wrap TracedRequestChannel to ensure requests have tracing headers.
+            chan = new TracedRequestChannel(chan);
+            chan = new TracedChannel(chan, "Dialogue-http-request");
+            LimitedChannel limitedChan = new ChannelToLimitedChannelAdapter(chan);
+            limitedChan = concurrencyLimiter(limitedChan, conf, clientMetrics, clock);
```

Comment:
a future piece of work would be to copy the state of the concurrencylimiter too. didn't want to go too crazy in this PR though.